challenges/product/challenge-1.md

```diff
@@ -0,0 +1,156 @@
+# Product Challenge
+
+## Rationale
+
+This exercise has two goals:
+
+It helps us understand what to expect from you as a technical leader, and
+whether you have the technical aptitude to engage with highly technical
+customers. It gives you a feel for what it’s like to work at Teleport. The
+exercise simulates day-to-day work, helping you better understand the team and
+our hiring process.
+
+We believe this approach is not only more effective, but also more enjoyable
+than the whiteboard or quiz-style interviews common in the industry.
+
+We appreciate your time and look forward to hacking on this little project
+together.
+
+
+## Objective
+
+Create a design document and implement a service that will let users sign on
+through SSO IDP of of any kind and get a WireGuard public key set up on their
+computer to be authorized to access a WireGuard network.
```

Comment:
I think this is too vague. It's clearly modeled after Tailscale, but Tailscale is doing a lot more than WireGuard (mesh networking, NAT traversal, etc). 

Pure WireGuard is simply allowing two peers that have exchanged public keys to communicate with each other, assuming there is already network connectivity between hosts.

In this challenge, we ask candidates to build a CLI - but what is the peer? What are they supposed to talk to?

My opinion - logging in via SSO should not be to get your own WireGuard keypair - anyone can generate that. Logging in via SSO should allow you to receive the public key of a peer (and provide your public key to that peer).

Reply:
Yes, that's what I meant in this challenge - you SSO, and as a part of SSO exchange, you exchange keys with one peer that serves nginx (or any other web service). To verify the result, we curl the service and succeed after logging in.